### PR TITLE
Carli/2242 ruler image matching bug

### LINKED ISSUE #2242: ruler annotation is not rendered properly on spatially matched images
**Describe the bug**
Found by @crocka 

If there are spatially matched images and if we create a ruler annotation, we will see the ruler annotation is not rendered properly on the matched images. See below as a demo.

https://github.com/CARTAvis/carta-frontend/assets/20819712/881bc77e-a7e8-4330-9a3f-f5a39484656e

**To Reproduce**
1. load two images and match them
2. create a ruler annotation and observe how it is rendered on both images

**Expected behavior**
the ruler annotation on the matched images should be rendered correctly as well

**Platform info (please complete the following information):**
 - OS [e.g. macOS Monterey]: macOS ventura 
 - Browser [e.g. chrome, safari, electron app]: chrome
 - Browser version [e.g. 22]: 106
 - Backend branch [e.g. dev, v3b2 release]: dev
 - Frontend branch [e.g. dev, v3b2 release]: dev

### PR DESCRIPTION
**Pull request by crocka**
**Description**

This PR fixes #2242. This bug is introduced by PR #2190, where the ``spatialTransformAst`` frameSet object in ``FrameStore`` was replace by a cmpMap. The fix is simply to introduce the frame of the reference image to the AST function.

**Checklist**

For linked issues (if there are):
- [x] assignee and label added
- [x] ZenHub issue connection, board status, and estimate updated

For the pull request:
- [x] reviewers and assignee added
- [x] ZenHub estimate, milestone, and release (if needed) added
- [x] e2e test passing / ~corresponding fix added~ / ~new e2e test created~
- [x] changelog updated / ~~no changelog update needed~~
- [x] ~~protobuf updated to the latest dev commit~~ / no protobuf update needed
- [x] `BackendService` unchanged / ~~`BackendService` changed and corresponding ICD test fix added~~

### CODE REVIEW

**Comment by kswang1029** [comment]
please proceed code review. A set of e2e tests will be added to address spatial matching of annotations.

**Review by loveluthien — APPROVED** [review]
The bug is fixed.

File: src/stores/Frame/AnnotationStore.ts [hunk 1]
```diff
@@ -334,7 +334,7 @@ export class CompassAnnotationStore extends RegionStore {
         this.modifiedTimestamp = performance.now();
     };
 
-    public getRegionApproximation(astTransform: AST.FrameSet, spatiallyMatched?: boolean, spatialTransform?: AST.Mapping): {northApproximatePoints: number[]; eastApproximatePoints: number[]} {
+    public getCompassApproximation(astTransform: AST.FrameSet, spatiallyMatched?: boolean, spatialTransform?: AST.Mapping): {northApproximatePoints: number[]; eastApproximatePoints: number[]} {
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/AnnotationStore.ts]: I suggest we rename `astTransform` to `wcsInfo`.

File: src/stores/Frame/AnnotationStore.ts [hunk 2]
```diff
@@ -489,22 +489,24 @@ export class RulerAnnotationStore extends RegionStore {
         this.modifiedTimestamp = performance.now();
     };
 
-    public getRegionApproximation(astTransform: AST.FrameSet | AST.Mapping, spatiallyMatched?: boolean): {xApproximatePoints: number[]; yApproximatePoints: number[]; hypotenuseApproximatePoints: number[]} {
+    public getCurveApproximation(astTransform: AST.FrameSet, mapping?: AST.Mapping): {xApproximatePoints: number[]; yApproximatePoints: number[]; hypotenuseApproximatePoints: number[]} {
```
Comment (YuHsuan-Hwang) [inline src/stores/Frame/AnnotationStore.ts]: I suggest we rename `astTransform` to `wcsInfo`.

File: wasm_src/ast_wrapper/ast_wrapper.cc [hunk 3]
```diff
@@ -444,7 +444,9 @@ EMSCRIPTEN_KEEPALIVE int pointList(AstFrameSet* wcsinfo, int npoint, double xin[
     double start[] = {xin[0], yin[0]};
     double finish[] = {xin[1], yin[1]};
 
-    double dist = astDistance(wcsinfo, start, finish);
+    AstFrame* frame = static_cast<AstFrame*>astGetFrame(wcsinfo, AST__CURRENT);
+
+    double dist = astDistance(frame, start, finish);
```
Comment (YuHsuan-Hwang) [inline wasm_src/ast_wrapper/ast_wrapper.cc]: Could we directly use `wcsinfo`? (same for `astOffset` below)

File: wasm_src/ast_wrapper/post.ts [hunk 4]
```diff
@@ -231,6 +231,7 @@ Module.transform3DPointArrays = function (wcsInfo: number, xIn: Float64Array, yI
 
 Module.getGeodesicPointArray = function (wcsInfo: number, npoint: number, start: {x: number, y: number}, finish: {x: number, y: number}) {
     // Return empty array if arguments are invalid
+
```
Comment (YuHsuan-Hwang) [inline wasm_src/ast_wrapper/post.ts]: Remove empty line.

**Review by YuHsuan-Hwang — CHANGES_REQUESTED** [review]
The new fix looks good! I have some minor suggestions. Please also update the change log, thanks.

File: CHANGELOG.md [hunk 5]
```diff
@@ -43,6 +43,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 * Fixed image arithmetic bug when there are square brackets in the expression ([#2220](https://github.com/CARTAvis/carta-frontend/issues/2220)).
 * Fixed missing tiles and distorted tiles during and after animation. ([#1988](https://github.com/CARTAvis/carta-frontend/issues/1988)).
 * Fixed the stuck of stokes animation at low frame rates ([#1293](https://github.com/CARTAvis/carta-backend/issues/1293)).
+* Fixed ruler annotation matching bug ([#2242](https://github.com/CARTAvis/carta-frontend/issues/2242)).
```
Comment (YuHsuan-Hwang) [inline CHANGELOG.md]: @crocka Please update to the latest dev and move it to [Unreleased].

**Review by YuHsuan-Hwang — APPROVED** [review]

**Review by kswang1029 — APPROVED** [review]

**Comment by kswang1029** [comment]
@loveluthien please have a final test (due to new code changes).

**Review by loveluthien — APPROVED** [review]
Tested.

File: src/stores/Frame/Region/RegionStore.ts [hunk 6]
```diff
@@ -466,7 +466,7 @@ export class RegionStore {
     };
 
     @action setCenter = (p: Point2D, skipUpdate = false) => {
-        if (this.regionType === CARTA.RegionType.LINE || this.regionType === CARTA.RegionType.ANNLINE || this.regionType === CARTA.RegionType.ANNVECTOR) {
+        if (this.regionType === CARTA.RegionType.LINE || this.regionType === CARTA.RegionType.ANNLINE || this.regionType === CARTA.RegionType.ANNVECTOR || this.regionType === CARTA.RegionType.ANNRULER) {
```
Comment (crocka) [inline src/stores/Frame/Region/RegionStore.ts:469]: @confluence Hi Adriana, could you please confirm if this PR fixed issue #2246?

**Pull request merged**